### PR TITLE
New parameter for home directory

### PR DESCRIPTION
**Pull request by joernott**
This solves elastic/puppet-logstash#367

### CODE REVIEW

**Comment by jarpy** [comment]
Thanks! Could you please add a section to the README so that people know it's there and how to use it without reading the code?

It will also need a test like [this one](https://github.com/elastic/puppet-logstash/blob/f3b9593af0f563eb4dd7a04c32692c4a57080870/spec/acceptance/class_logstash_spec.rb#L106-L113) before we can merge. I can work on the test next week if you don't get to it before then.

Cheers.

**Comment by joernott** [comment]
Adding to the readme is rather trivial. Writing the test is quite a challenge, as this parameter is not used in the class directly but in the sub classes.

When running tests on our test platform, I encounbtered another issue with floating classes (trying to install the package before the yum repo was available), so I switched from "include" to "contain".

After that, I will try to see, what my test runs write into the logs, so I can write an appropriate test.

**Comment by fatmcgav** [comment]
@joernott With regards to writing an acceptance test, it should be possible to use the `user` server-spec resource:
```
have_home_directory
In order to test a user have a given home directory, you should use have_home_directory matcher.

describe user('root') do
  it { should have_home_directory '/root' }
end
```
- https://serverspec.org/resource_types.html#user

**Comment by jarpy** [comment]
Is this patch intended to be used with Logstash installations that are provided by other means (ie. not by this module)? I guess my question is: "_how is Logstash getting into a custom location in the first place_?".

If I read this correctly, setting the `$home_dir` won't actually change were the module installs Logstash, since that is defined by the deb/rpm packages.

**Comment by joernott** [comment]
I am aware of that. Currently, we package our own logstash (and the rest of the stack to achieve compliance to our folder structure rules. We have an ugly wrapper around the puppet-logstash module which creates symlinks and has dependencies on resources inside the puppet-logstash module to work around puppet-logstash using a hard coded file path to determine file locations which are only valid if you use the original .deb|.rpm packages. Even if you unpack the official tar.gz, this module won't work. The puppet-elasticsearch module has a variable $home_dir which they use instead of hard coded paths for that reason. So this PR just introduces that feature.

**Comment by joernott** [comment]
Rebase done

**Review by ekohl — CHANGES_REQUESTED** [review]
The tests didn't run, but I'd expect them to fail. The declaration of `$home_dir` on line 161 looks invalid to me.

File: manifests/init.pp [hunk 1]
```diff
@@ -167,7 +171,11 @@
...
-  include logstash::package
-  include logstash::config
-  include logstash::service
+  contain logstash::package
+  contain logstash::config
+  contain logstash::service
+
+  Class['logstash::package']
+  -> Class['logstash::config']
+  -> Class['logstash::service']
```
[the remark this answers is not in the document]
Comment (joernott) [reply manifests/init.pp:180]: I guess, that is my current customers design pattern kicking in. I will revert that part

**Review by ekohl — COMMENTED** [review]
I still think you need to remove this line:
https://github.com/voxpupuli/puppet-logstash/blob/a293c272e16cc0fad5c6f3bea2fc6c32e69c0f00/manifests/init.pp#L157

**Review by ekohl — APPROVED** [review]

**Pull request merged**